File: src/algorithms/cohenSutherland.ts
import { ClippingRect, CohenSutherlandStepState, Point2D } from '../types';

export const INSIDE = 0; // 0000
export const LEFT = 1;   // 0001
export const RIGHT = 2;  // 0010
export const BOTTOM = 4; // 0100
export const TOP = 8;    // 1000

export function computeOutCode(x: number, y: number, clip: ClippingRect): number {
  let code = INSIDE;
  if (x < clip.xmin) {
    code |= LEFT;
  } else if (x > clip.xmax) {
    code |= RIGHT;
  }
  if (y < clip.ymin) {
    code |= BOTTOM;
  } else if (y > clip.ymax) {
    code |= TOP;
  }
  return code;
}

export function toBinary4(code: number): string {
  return (code >>> 0).toString(2).padStart(4, '0');
}

export function generateCohenSutherlandSteps(
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  clip: ClippingRect
): CohenSutherlandStepState[] {
  const steps: CohenSutherlandStepState[] = [];
  const originalLine = { p1: { x: x1, y: y1 }, p2: { x: x2, y: y2 } };
  const rejectedSegments: Array<{ p1: Point2D; p2: Point2D }> = [];

  let curP1: Point2D = { x: x1, y: y1 };
  let curP2: Point2D = { x: x2, y: y2 };
  let code1 = computeOutCode(curP1.x, curP1.y, clip);
  let code2 = computeOutCode(curP2.x, curP2.y, clip);

  let stepIdx = 0;
  let done = false;
  let maxIters = 8;

  // Initial step 0: Evaluation
  steps.push({
    step: 0,
    totalSteps: 1, // dynamically updated at end
    p1: { ...curP1 },
    p2: { ...curP2 },
    code1: toBinary4(code1),
    code2: toBinary4(code2),
    code1Val: code1,
    code2Val: code2,
    action: 'EVALUATE',
    explanation: `Initial line endpoints: P1(${curP1.x}, ${curP1.y}) code = ${toBinary4(code1)}, P2(${curP2.x}, ${curP2.y}) code = ${toBinary4(code2)}. Clipping window: [${clip.xmin}, ${clip.xmax}] × [${clip.ymin}, ${clip.ymax}].`,
    originalLine,
    clippingWindow: clip,
    rejectedSegments: [...rejectedSegments],
    acceptedSegment: null
  });

  while (!done && maxIters > 0) {
    maxIters--;
    stepIdx++;

    // Check trivial accept
    if ((code1 | code2) === 0) {
      done = true;
      steps.push({
        step: stepIdx,
        totalSteps: stepIdx,
        p1: { ...curP1 },
        p2: { ...curP2 },
        code1: toBinary4(code1),
        code2: toBinary4(code2),
        code1Val: code1,
        code2Val: code2,
        action: 'ACCEPT',
        explanation: `Trivial ACCEPT! (code1 | code2) == 0 (both endpoints inside window). Visible line segment is from (${curP1.x.toFixed(2)}, ${curP1.y.toFixed(2)}) to (${curP2.x.toFixed(2)}, ${curP2.y.toFixed(2)}).`,
        originalLine,
        clippingWindow: clip,
        rejectedSegments: [...rejectedSegments],
        acceptedSegment: { p1: { ...curP1 }, p2: { ...curP2 } }
      });
      break;
    }

    // Check trivial reject
    if ((code1 & code2) !== 0) {
      done = true;
      rejectedSegments.push({ p1: { ...curP1 }, p2: { ...curP2 } });
      steps.push({
        step: stepIdx,
        totalSteps: stepIdx,
        p1: { ...curP1 },
        p2: { ...curP2 },
        code1: toBinary4(code1),
        code2: toBinary4(code2),
        code1Val: code1,
        code2Val: code2,
        action: 'REJECT',
        explanation: `Trivial REJECT! (code1 & code2) = ${toBinary4(code1 & code2)} ≠ 0. Both endpoints share an outside half-plane. Line is completely outside the clipping window.`,
        originalLine,
        clippingWindow: clip,
        rejectedSegments: [...rejectedSegments],
        acceptedSegment: null
      });
      break;
    }

    // Clipping is necessary
    const codeOut = code1 !== 0 ? code1 : code2;
    const isP1 = code1 !== 0;
    let ix = 0;
    let iy = 0;
    let boundary: 'LEFT' | 'RIGHT' | 'BOTTOM' | 'TOP' = 'TOP';

    const dx = curP2.x - curP1.x;
    const dy = curP2.y - curP1.y;

    if (codeOut & TOP) {
      boundary = 'TOP';
      iy = clip.ymax;
      ix = curP1.x + (dx * (clip.ymax - curP1.y)) / dy;
    } else if (codeOut & BOTTOM) {
      boundary = 'BOTTOM';
      iy = clip.ymin;
      ix = curP1.x + (dx * (clip.ymin - curP1.y)) / dy;
    } else if (codeOut & RIGHT) {
      boundary = 'RIGHT';
      ix = clip.xmax;
      iy = curP1.y + (dy * (clip.xmax - curP1.x)) / dx;
    } else if (codeOut & LEFT) {
      boundary = 'LEFT';
      ix = clip.xmin;
      iy = curP1.y + (dy * (clip.xmin - curP1.x)) / dx;
    }

    const oldPt = isP1 ? { ...curP1 } : { ...curP2 };
    const newPt: Point2D = { x: Number(ix.toFixed(2)), y: Number(iy.toFixed(2)) };
    rejectedSegments.push({ p1: oldPt, p2: newPt });

    if (isP1) {
      curP1 = newPt;
      code1 = computeOutCode(curP1.x, curP1.y, clip);
    } else {
      curP2 = newPt;
      code2 = computeOutCode(curP2.x, curP2.y, clip);
    }

    steps.push({
      step: stepIdx,
      totalSteps: stepIdx + 1,
      p1: { ...curP1 },
      p2: { ...curP2 },
      code1: toBinary4(code1),
      code2: toBinary4(code2),
      code1Val: code1,
      code2Val: code2,
      action: 'CLIP',
      clippedPointIndex: isP1 ? 1 : 2,
      intersectionPoint: newPt,
      boundaryTested: boundary,
      explanation: `Line clipped against ${boundary} boundary! P${isP1 ? 1 : 2} (${oldPt.x}, ${oldPt.y}) replaced with intersection point (${newPt.x}, ${newPt.y}). New code: ${toBinary4(isP1 ? code1 : code2)}.`,
      originalLine,
      clippingWindow: clip,
      rejectedSegments: [...rejectedSegments],
      acceptedSegment: null
    });
  }

  // Set totalSteps correctly on all steps
  const totalSteps = steps.length - 1;
  steps.forEach(s => s.totalSteps = totalSteps);

  return steps;
}

File: python_project/algorithms/cohen_sutherland.py
"""
Cohen-Sutherland Line Clipping Algorithm with 4-bit region outcodes.
"""

INSIDE = 0  # 0000
LEFT = 1    # 0001
RIGHT = 2   # 0010
BOTTOM = 4  # 0100
TOP = 8     # 1000

def compute_outcode(x, y, xmin, ymin, xmax, ymax):
    code = INSIDE
    if x < xmin:
        code |= LEFT
    elif x > xmax:
        code |= RIGHT
    if y < ymin:
        code |= BOTTOM
    elif y > ymax:
        code |= TOP
    return code

def to_binary4(code):
    return f"{code:04b}"

class CohenSutherland:
    def __init__(self, x1, y1, x2, y2, xmin, ymin, xmax, ymax):
        self.x1 = float(x1)
        self.y1 = float(y1)
        self.x2 = float(x2)
        self.y2 = float(y2)
        self.xmin = float(xmin)
        self.ymin = float(ymin)
        self.xmax = float(xmax)
        self.ymax = float(ymax)

    def generate_steps(self):
        steps = []
        cur_x1, cur_y1 = self.x1, self.y1
        cur_x2, cur_y2 = self.x2, self.y2

        code1 = compute_outcode(cur_x1, cur_y1, self.xmin, self.ymin, self.xmax, self.ymax)
        code2 = compute_outcode(cur_x2, cur_y2, self.xmin, self.ymin, self.xmax, self.ymax)

        step_idx = 0
        original_line = ((self.x1, self.y1), (self.x2, self.y2))
        window = (self.xmin, self.ymin, self.xmax, self.ymax)
        rejected_segments = []

        steps.append({
            "step": 0,
            "p1": (cur_x1, cur_y1),
            "p2": (cur_x2, cur_y2),
            "code1": to_binary4(code1),
            "code2": to_binary4(code2),
            "action": "EVALUATE",
            "explanation": f"Initial line P1({cur_x1}, {cur_y1}) [{to_binary4(code1)}] to P2({cur_x2}, {cur_y2}) [{to_binary4(code2)}]. Window: [{self.xmin}, {self.xmax}] x [{self.ymin}, {self.ymax}].",
            "original_line": original_line,
            "window": window,
            "accepted_segment": None,
            "rejected_segments": list(rejected_segments)
        })

        done = False
        max_iters = 8

        while not done and max_iters > 0:
            max_iters -= 1
            step_idx += 1

            if (code1 | code2) == 0:
                # Trivial accept
                done = True
                steps.append({
                    "step": step_idx,
                    "p1": (cur_x1, cur_y1),
                    "p2": (cur_x2, cur_y2),
                    "code1": to_binary4(code1),
                    "code2": to_binary4(code2),
                    "action": "ACCEPT",
                    "explanation": f"Trivial ACCEPT: Both codes are 0000. Line segment from ({cur_x1:.2f}, {cur_y1:.2f}) to ({cur_x2:.2f}, {cur_y2:.2f}) is visible.",
                    "original_line": original_line,
                    "window": window,
                    "accepted_segment": ((cur_x1, cur_y1), (cur_x2, cur_y2)),
                    "rejected_segments": list(rejected_segments)
                })
                break

            if (code1 & code2) != 0:
                # Trivial reject
                done = True
                rejected_segments.append(((cur_x1, cur_y1), (cur_x2, cur_y2)))
                steps.append({
                    "step": step_idx,
                    "p1": (cur_x1, cur_y1),
                    "p2": (cur_x2, cur_y2),
                    "code1": to_binary4(code1),
                    "code2": to_binary4(code2),
                    "action": "REJECT",
                    "explanation": f"Trivial REJECT: Logical AND ({to_binary4(code1 & code2)}) != 0. Both points share an exterior region.",
                    "original_line": original_line,
                    "window": window,
                    "accepted_segment": None,
                    "rejected_segments": list(rejected_segments)
                })
                break

            # Clip against boundary
            code_out = code1 if code1 != 0 else code2
            is_p1 = (code1 != 0)
            dx = cur_x2 - cur_x1
            dy = cur_y2 - cur_y1

            if code_out & TOP:
                boundary = "TOP"
                iy = self.ymax
                ix = cur_x1 + (dx * (self.ymax - cur_y1)) / dy
            elif code_out & BOTTOM:
                boundary = "BOTTOM"
                iy = self.ymin
                ix = cur_x1 + (dx * (self.ymin - cur_y1)) / dy
            elif code_out & RIGHT:
                boundary = "RIGHT"
                ix = self.xmax
                iy = cur_y1 + (dy * (self.xmax - cur_x1)) / dx
            elif code_out & LEFT:
                boundary = "LEFT"
                ix = self.xmin
                iy = cur_y1 + (dy * (self.xmin - cur_x1)) / dx

            old_pt = (cur_x1, cur_y1) if is_p1 else (cur_x2, cur_y2)
            new_pt = (round(ix, 2), round(iy, 2))
            rejected_segments.append((old_pt, new_pt))

            if is_p1:
                cur_x1, cur_y1 = new_pt
                code1 = compute_outcode(cur_x1, cur_y1, self.xmin, self.ymin, self.xmax, self.ymax)
            else:
                cur_x2, cur_y2 = new_pt
                code2 = compute_outcode(cur_x2, cur_y2, self.xmin, self.ymin, self.xmax, self.ymax)

            steps.append({
                "step": step_idx,
                "p1": (cur_x1, cur_y1),
                "p2": (cur_x2, cur_y2),
                "code1": to_binary4(code1),
                "code2": to_binary4(code2),
                "action": "CLIP",
                "boundary": boundary,
                "intersection": new_pt,
                "explanation": f"Clipped against {boundary} boundary. P{1 if is_p1 else 2} moved to intersection ({new_pt[0]}, {new_pt[1]}). New code: {to_binary4(code1 if is_p1 else code2)}.",
                "original_line": original_line,
                "window": window,
                "accepted_segment": None,
                "rejected_segments": list(rejected_segments)
            })

        total = len(steps) - 1
        for s in steps:
            s["total_steps"] = total
        return steps
